### DISCUSSION TITLE
add new colum to table elements

### DISCUSSION THREAD

**Discussion by Jaztty** [iTwin/itwinjs-core#6685] [Q&A]
![ImgTableElement](https://github.com/iTwin/itwinjs-core/assets/89955230/74cc53c3-e390-4154-9ad2-1269dc1e7db8)

Hello, currently in the elements table I see several columns such as
CodeValue, UserLabel, FederationGuid, JsonProperties ... and others more...

It would be possible to add a new column called "NewColumn" ? 

If possible, some recommendation to develop it (code, functions or libraries). I am new to the community and I would greatly appreciate your help! Greetings.

**Comment by ColinKerr** [comment]
Are you just trying to add a new column in the table UI with data provided from some external source or do you want to add a new property whose values will be stored in the iModel?

**Reply by Jaztty** [reply]
I want to add a new column in the imodel and also have it displayed in the viewer. as is the case of the UserLabel column that I show in the photos. In my case the new column I want to add would be just a string.
![imodeltableUserLabel](https://github.com/iTwin/itwinjs-core/assets/89955230/2142d2d1-b8bc-4120-9c4e-118347ba23e0)
![viewuserlabel](https://github.com/iTwin/itwinjs-core/assets/89955230/1d94aedd-52e6-4e86-b13e-f4c940c645c7)

**Reply by Jaztty** [reply]
@ColinKerr Reading your question, I think I know what you mean about external sources. I mainly want to achieve that and then make those modifications go to the imodel. Is it possible?

**Reply by ColinKerr** [reply]
@Jaztty if you want to import new data into an iModel there are several options.  What you choose depends on the source of the data and if it's going to be updated again from the original source.  

If the data is coming in via a connector from an external source (e.g. a dgn, dwg or rvt file) then the easiest thing to do is to update the source file and run the connector again.  This synchronization will bring in the new data from the source file automatically.  

If the iModel is no longer being updated from the source files so you can modify the data directly in the iModel by adding a new property to the class then use the element API to update the element with values for the new data.

Update schemas: https://www.itwinjs.org/learning/serializing-xml-schemas/#schemas-from-an-imodel

> It is also possible to edit schemas as xml or JSON using the xsd or json schemas found here: https://github.com/iTwin/bis-schemas/tree/master/Standard

Update Element: https://www.itwinjs.org/reference/core-backend/imodels/imodeldb/imodeldb.elements/updateelement/

> For this to you work you need to have the iModel [open read/write](https://www.itwinjs.org/reference/core-backend/imodels/briefcasedb/openstatic/) and there are no locks.  To learn more about locks see: https://www.itwinjs.org/learning/backend/concurrencycontrol/

If the data is being continuously updated by the connector it is harder to augment because the connector may remove any change you make on the next sync.  I'll just point you to some info on writing your own connector here, but more discussion is necessary if this is your use case.  https://www.itwinjs.org/learning/writeaconnector/

**Reply by Jaztty** [reply]
@ColinKerr Hello,

Thank you for your guindance! I have been working on modifying the schema based on the documentation you provided (Serializing XML Schemas) and the guidelines in your previous messages. My understanding is that I need to generate a new schema starting from an existing class, such as biscore.physicalelement, and then add properties to this new schema copy.

I initially tried to extend biscore.element, but encountered a sealed property that prevented me from doing so. You did not mention adding a new schema as an extension in your comments, so I am unsure if my interpretation aligns with your instructions.

Here is what I have tried so far:

1- Creating a new schema based on PhysicalElement: 


```
const loader = new SchemaLoader(name => iModelDb.getSchemaProps(name));
const editor = new SchemaContextEditor(loader.context);
 const createSchemaResult = await editor.createSchema("ExtraDirection", "ED", 1, 0, 0);
if (createSchemaResult.errorMessage) {
   throw new Error("Failed to create schema: " + createSchemaResult.errorMessage);
}

const bisSchema = loader.getSchema("BisCore");
const addRefResult = await editor.addSchemaReference(createSchemaResult.schemaKey, bisSchema);
if (addRefResult.errorMessage) {
   throw new Error("Failed to add reference schema: " + addRefResult.errorMessage);
}

const createClassResult = await editor.entities.createElement(createSchemaResult.schemaKey, "PinW", ECClassModifier.None, bisSchema.getSchemaItemKey("BisCore.PhysicalElement"));
if (createClassResult.errorMessage) {
   throw new Error("Failed to create class: " + createClassResult.errorMessage);
}

const createPropResult = await editor.entities.createProperty(createClassResult.itemKey, "Direction", PrimitiveType.Double, "PinWD");
if (createPropResult.errorMessage) {
   throw new Error("Failed to create property: " + createPropResult.errorMessage);
}

const placeSchema = await editor.getSchema(createSchemaResult.schemaKey);
if (!placeSchema) {
   throw new Error("Schema Key not found in context");
}
const schemaXml = await SchemaXml.writeString(placeSchema);
await iModelDb.importSchemaStrings([schemaXml]);
``` 

2- Failed attempt to add a property directly to biscore.element:

```
const bisSchema = loader.getSchema("BisCore");
const schemeKey = bisSchema.getSchemaItemKey("BisCore.Element");

const createPropResult = await editor.entities.createProperty(schemeKey, "Direction", PrimitiveType.String, "PinWD");
if (createPropResult.errorMessage) {
   throw new Error("Failed to create property: " + createPropResult.errorMessage);
}

const placeSchema = bisSchema;
const schemaXml = await SchemaXml.writeString(placeSchema);
await iModelDb.importSchemaStrings([schemaXml]);
```

I think I may be doing something wrong in the second case, or perhaps it's not feasible to modify biscore.element directly.

Could you confirm if my approach in the first example is correct for creating and extending schemas in iModel.js? Additionally, should I be using a method like insertElement instead of updateElement for this new class since it would start empty and needs to be associated with an existing element?

**Comment by Jaztty** [comment]
Hello, I have already managed to add the column into the table, now I would like to be able to show that column in the viewer. What class or library should I use?
@ColinKerr

**Comment by Jaztty** [comment]
@pmconne this would be possible?
